In [5]:
# import required modules
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
#from scripts import handler, db_connect
import sys
import os

In [6]:
# Get the absolute path to the parent directory
parent_dir = os.path.dirname(os.getcwd())

# Add the parent directory to sys.path
sys.path.append(parent_dir)
from scripts import handler, db_connect

In [7]:
db_url = 'postgresql://root:root@localhost:5432/tellco'
conect = db_connect.PostgresDBConnector(db_url)

In [8]:
tell_df = conect.query_to_dataframe()
tell_df.head()

Bearer Id            Start  Start ms              End  End ms  \
0  1.311448e+19   4/4/2019 12:01     770.0  4/25/2019 14:35   662.0   
1  1.311448e+19   4/9/2019 13:04     235.0   4/25/2019 8:15   606.0   
2  1.311448e+19   4/9/2019 17:42       1.0  4/25/2019 11:58   652.0   
3  1.311448e+19   4/10/2019 0:31     486.0   4/25/2019 7:36   171.0   
4  1.311448e+19  4/12/2019 20:10     565.0  4/25/2019 10:40   954.0   

   Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0  1823652.0  2.082014e+14   3.366496e+10  3.552121e+13   
1  1365104.0  2.082019e+14   3.368185e+10  3.579401e+13   
2  1361762.0  2.082003e+14   3.376063e+10  3.528151e+13   
3  1321509.0  2.082014e+14   3.375034e+10  3.535661e+13   
4  1089009.0  2.082014e+14   3.369980e+10  3.540701e+13   

      Last Location Name  ...  Youtube DL (Bytes)  Youtube UL (Bytes)  \
0  9.16456699548519E+015  ...          15854611.0           2501332.0   
1                L77566A  ...          20247395.0          19111729.0   
2                D42335A  ...          19725661.0          14699576.0   
3                T21824A  ...          21388122.0          15146643.0   
4                D88865A  ...          15259380.0          18962873.0   

   Netflix DL (Bytes)  Netflix UL (Bytes)  Gaming DL (Bytes)  \
0           8198936.0           9656251.0        278082303.0   
1          18338413.0          17227132.0        608750074.0   
2          17587794.0           6163408.0        229584621.0   
3          13994646.0           1097942.0        799538153.0   
4          17124581.0            415218.0        527707248.0   

   Gaming UL (Bytes)  Other DL (Bytes)  Other UL (Bytes)  Total UL (Bytes)  \
0         14344150.0       171744450.0         8814393.0        36749741.0   
1          1170709.0       526904238.0        15055145.0        53800391.0   
2           395630.0       410692588.0         4215763.0        27883638.0   
3         10849722.0       749039933.0        12797283.0        43324218.0   
4          3529801.0       550709500.0        13910322.0        38542814.0   

   Total DL (Bytes)  
0       308879636.0  
1       653384965.0  
2       279807335.0  
3       846028530.0  
4       569138589.0  

[5 rows x 55 columns]

In [10]:
df = handler.MissingValueHandler(tell_df)
df.df_info()


DataFrame Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150001 entries, 0 to 150000
Data columns (total 55 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   Bearer Id                                 149010 non-null  float64
 1   Start                                     150000 non-null  object 
 2   Start ms                                  150000 non-null  float64
 3   End                                       150000 non-null  object 
 4   End ms                                    150000 non-null  float64
 5   Dur. (ms)                                 150000 non-null  float64
 6   IMSI                                      149431 non-null  float64
 7   MSISDN/Number                             148935 non-null  float64
 8   IMEI                                      149429 non-null  float64
 9   Last Location Name                        148848 non-null  object 
 

In [11]:
df.check_duplicate()


Duplicated values:
0 duplicated rows


In [12]:
df.calculate_missing_percentage()

Bearer Id                                    0.660662
Start                                        0.000667
Start ms                                     0.000667
End                                          0.000667
End ms                                       0.000667
Dur. (ms)                                    0.000667
IMSI                                         0.379997
MSISDN/Number                                0.710662
IMEI                                         0.381331
Last Location Name                           0.768662
Avg RTT DL (ms)                             18.552543
Avg RTT UL (ms)                             18.541210
Avg Bearer TP DL (kbps)                      0.000667
Avg Bearer TP UL (kbps)                      0.000667
TCP DL Retrans. Vol (Bytes)                 58.763608
TCP UL Retrans. Vol (Bytes)                 64.432237
DL TP < 50 Kbps (%)                          0.502663
50 Kbps < DL TP < 250 Kbps (%)               0.502663
250 Kbps < DL TP < 1 Mbps (%

In [35]:
#df.handle_missing_values()
for col in tell_df.columns:
    if tell_df[col].dtype == np.float64:
        # Fill missing values with the mean for numeric columns
        tell_df[col] = tell_df[col].fillna(tell_df[col].mean())
        
    else:
        # Leave object columns as they are
        pass

In [37]:
tell_df['Start'] = pd.to_datetime(tell_df['Start'])
tell_df['End'] = pd.to_datetime(tell_df['End'])

In [40]:
tell_df.dropna(inplace=True)

In [46]:
df.calculate_missing_percentage()

Bearer Id                                   0.0
Start                                       0.0
Start ms                                    0.0
End                                         0.0
End ms                                      0.0
Dur. (ms)                                   0.0
IMSI                                        0.0
MSISDN/Number                               0.0
IMEI                                        0.0
Last Location Name                          0.0
Avg RTT DL (ms)                             0.0
Avg RTT UL (ms)                             0.0
Avg Bearer TP DL (kbps)                     0.0
Avg Bearer TP UL (kbps)                     0.0
TCP DL Retrans. Vol (Bytes)                 0.0
TCP UL Retrans. Vol (Bytes)                 0.0
DL TP < 50 Kbps (%)                         0.0
50 Kbps < DL TP < 250 Kbps (%)              0.0
250 Kbps < DL TP < 1 Mbps (%)               0.0
DL TP > 1 Mbps (%)                          0.0
UL TP < 10 Kbps (%)                     

# Non-Graphical Univariate Analysis
- Objective: To understand the variability within each quantitative variable.

- Importance: This step helps in identifying outliers, understanding the spread of data, and assessing the normality of distributions. It's crucial for setting baselines and understanding the scale of each variable, which is essential for subsequent analyses and for making informed decisions about data preprocessing.

- Method: Compute measures like mean, median, standard deviation, minimum, and maximum values for each variable. Interpret these values to understand the central tendency and dispersion of the data.

In [45]:
tell_df.describe()

Bearer Id                          Start       Start ms  \
count  1.488370e+05                         148837  148837.000000   
mean   1.013886e+19  2019-04-26 12:40:53.477025280     499.163743   
min    6.917538e+18            2019-04-04 12:01:00       0.000000   
25%    7.349883e+18            2019-04-25 00:46:00     250.000000   
50%    7.349883e+18            2019-04-26 08:52:00     499.000000   
75%    1.304243e+19            2019-04-27 18:30:00     749.000000   
max    1.318654e+19            2019-04-29 07:28:00     999.000000   
std    2.893144e+18                            NaN     288.621038   

                                 End         End ms     Dur. (ms)  \
count                         148837  148837.000000  1.488370e+05   
mean   2019-04-27 17:51:10.124498432     498.678642  1.050161e+05   
min              2019-04-24 22:59:00       0.000000  7.142000e+03   
25%              2019-04-26 07:35:00     251.000000  5.815100e+04   
50%              2019-04-27 23:04:00     500.000000  8.639900e+04   
75%              2019-04-29 07:20:00     750.000000  1.328480e+05   
max              2019-04-30 23:59:00     999.000000  1.859336e+06   
std                              NaN     288.100318  8.101183e+04   

               IMSI  MSISDN/Number          IMEI  Avg RTT DL (ms)  ...  \
count  1.488370e+05   1.488370e+05  1.488370e+05    148837.000000  ...   
mean   2.082016e+14   4.191531e+10  4.847882e+13       106.584605  ...   
min    2.040471e+14   3.360100e+10  4.400152e+11         0.000000  ...   
25%    2.082014e+14   3.365140e+10  3.546071e+13        35.000000  ...   
50%    2.082015e+14   3.366375e+10  3.572211e+13        54.000000  ...   
75%    2.082018e+14   3.368410e+10  8.611970e+13       109.795706  ...   
max    2.140743e+14   8.823971e+14  9.900120e+13     96923.000000  ...   
std    2.153075e+10   2.448249e+12  2.241713e+13       527.781157  ...   

       Youtube DL (Bytes)  Youtube UL (Bytes)  Netflix DL (Bytes)  \
count        1.488370e+05        1.488370e+05        1.488370e+05   
mean         1.163360e+07        1.100944e+07        1.162611e+07   
min          5.300000e+01        1.050000e+02        4.200000e+01   
25%          5.833501e+06        5.517965e+06        5.777563e+06   
50%          1.161415e+07        1.101188e+07        1.164062e+07   
75%          1.744780e+07        1.651545e+07        1.747004e+07   
max          2.325910e+07        2.201196e+07        2.325919e+07   
std          6.711085e+06        6.345340e+06        6.725142e+06   

       Netflix UL (Bytes)  Gaming DL (Bytes)  Gaming UL (Bytes)  \
count        1.488370e+05       1.488370e+05       1.488370e+05   
mean         1.100259e+07       4.221243e+08       8.290760e+06   
min          3.500000e+01       2.516000e+03       5.900000e+01   
25%          5.476694e+06       2.106002e+08       4.133141e+06   
50%          1.099935e+07       4.235603e+08       8.293617e+06   
75%          1.650605e+07       6.332922e+08       1.243249e+07   
max          2.201196e+07       8.434419e+08       1.655879e+07   
std          6.358787e+06       2.439955e+08       4.781462e+06   

       Other DL (Bytes)  Other UL (Bytes)  Total UL (Bytes)  Total DL (Bytes)  
count      1.488370e+05      1.488370e+05      1.488370e+05      1.488370e+05  
mean       4.211534e+08      8.263305e+06      4.112291e+07      4.547221e+08  
min        3.290000e+03      1.480000e+02      2.866892e+06      7.114041e+06  
25%        2.103064e+08      4.144464e+06      3.322038e+07      2.432136e+08  
50%        4.218317e+08      8.266766e+06      4.114569e+07      4.560453e+08  
75%        6.317047e+08      1.238209e+07      4.903587e+07      6.657757e+08  
max        8.434425e+08      1.655882e+07      7.833131e+07      9.029696e+08  
std        2.431781e+08      4.768782e+06      1.127854e+07      2.441684e+08  

[8 rows x 52 columns]

### Interpreting Summary Statistics
- Mean Values: The mean values provide a general overview of the dataset. For example, the mean duration of 498.678642 milliseconds indicates that the average session lasts approximately half a second. The mean data transfer values for different applications suggest which applications are the most data-intensive.

- Min and Max Values: The minimum and maximum values highlight the range of the dataset. For instance, the minimum duration of 0 milliseconds suggests that some sessions are instantaneous, while the maximum duration of 999 milliseconds indicates the longest observed session. Similarly, the data transfer values show the smallest and largest amounts of data transferred by different applications.

- Percentiles (25%, 50%, 75%): These values offer insights into the distribution of the data. For example, the 25th percentile duration of 251 milliseconds indicates that 25% of sessions last less than this duration, while the 75th percentile duration of 750 milliseconds shows that 75% of sessions last less than this duration. This can help in understanding the typical session lengths and data transfer volumes.
- Standard Deviation: The standard deviation measures the variability of the data. A high standard deviation in duration or data transfer volumes suggests a wide variation in session lengths or data usage, which could indicate a need for more granular analysis or targeted interventions.
### Drawing Insights
Based on the summary statistics, several insights can be drawn:

- Session Duration: The average session duration is relatively short, suggesting that users engage with network services briefly. This could imply a high frequency of connections, which might impact network efficiency and user experience due to connection overhead.
- Data Transfer: The data transfer values for different applications indicate which services are the most data-intensive. For example, YouTube and Netflix have high data transfer volumes, suggesting that these services could benefit from optimized streaming protocols or bandwidth prioritization.
- Variability: The high standard deviation in data transfer volumes suggests that data usage varies widely among users. This variability could complicate network management and user experience optimization efforts.
- Network Optimization: Given the high data transfer volumes for certain applications and the variability in data usage, network operators might consider implementing Quality of Service (QoS) policies to prioritize critical services and manage bandwidth more efficiently. Additionally, analyzing the distribution of data usage over time could help in scheduling maintenance or upgrading network infrastructure during off-peak hours to minimize impact on user experience.


In summary, the insights derived from this non-graphical univariate analysis can guide network operators in optimizing network performance and enhancing user experience by addressing the most data-intensive applications, managing bandwidth more effectively, and considering the variability in data usage patterns.
Mean Values: The mean values provide a general overview of the dataset. For example, the mean duration of 498.678642 milliseconds indicates that the average session lasts approximately half a second. The mean data transfer values for different applications suggest which applications are the most data-intensive.

# Graphical Univariate Analysis
- Objective: To visually inspect the distribution and characteristics of each variable.

- Importance: Visualization aids in spotting trends, outliers, and anomalies that might not be apparent through numerical analysis alone. It's vital for exploratory data analysis and for making initial assessments about the data's quality and potential issues.

- Method: Plot histograms, box plots, and density plots for each variable. These plots can reveal skewness, multimodality, and the presence of outliers, which are critical for understanding the data's underlying structure.

In [47]:
tell_df.columns

Index(['Bearer Id', 'Start', 'Start ms', 'End', 'End ms', 'Dur. (ms)', 'IMSI',
       'MSISDN/Number', 'IMEI', 'Last Location Name', 'Avg RTT DL (ms)',
       'Avg RTT UL (ms)', 'Avg Bearer TP DL (kbps)', 'Avg Bearer TP UL (kbps)',
       'TCP DL Retrans. Vol (Bytes)', 'TCP UL Retrans. Vol (Bytes)',
       'DL TP < 50 Kbps (%)', '50 Kbps < DL TP < 250 Kbps (%)',
       '250 Kbps < DL TP < 1 Mbps (%)', 'DL TP > 1 Mbps (%)',
       'UL TP < 10 Kbps (%)', '10 Kbps < UL TP < 50 Kbps (%)',
       '50 Kbps < UL TP < 300 Kbps (%)', 'UL TP > 300 Kbps (%)',
       'HTTP DL (Bytes)', 'HTTP UL (Bytes)', 'Activity Duration DL (ms)',
       'Activity Duration UL (ms)', 'Dur. (ms).1', 'Handset Manufacturer',
       'Handset Type', 'Nb of sec with 125000B < Vol DL',
       'Nb of sec with 1250B < Vol UL < 6250B',
       'Nb of sec with 31250B < Vol DL < 125000B',
       'Nb of sec with 37500B < Vol UL',
       'Nb of sec with 6250B < Vol DL < 31250B',
       'Nb of sec with 6250B < Vol UL < 37500B',
